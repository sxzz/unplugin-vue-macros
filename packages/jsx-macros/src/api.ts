export * from './core'
export { restructure } from './core/define-component/restructure'
export { createPropsDefaultProxy } from './core/helper/with-defaults'
