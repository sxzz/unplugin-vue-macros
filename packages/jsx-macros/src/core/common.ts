import type { Program } from '@babel/types'
import type { MagicString } from 'vue/compiler-sfc'

const importedMap = new WeakMap<MagicString, Set<string>>()
export function importHelper({
  s,
  ast,
  local,
  as = local,
  isDefault,
  from = 'vue',
  offset = 0,
}: {
  s: MagicString
  ast: Program
  local: string
  as?: string
  isDefault?: boolean
  from?: string
  offset?: number
}): string {
  let hasBeenImported = false
  ast.body.forEach((node) => {
    if (node.type === 'ImportDeclaration' && node.source.value === from) {
      for (const specifier of node.specifiers) {
        if (specifier.local.name === as) {
          hasBeenImported = true
          break
        }
      }
    }
  })
  if (hasBeenImported) return as

  const imported = isDefault ? 'default' : local
  const cacheKey = `${from}@${imported}`
  if (!importedMap.get(s)?.has(cacheKey)) {
    s.appendLeft(
      offset,
      `\nimport ${
        isDefault ? as : `{ ${imported}${as !== local ? ` as ${as}` : ''} }`
      } from ${JSON.stringify(from)};`,
    )
    if (!importedMap.has(s)) {
      importedMap.set(s, new Set([cacheKey]))
    } else {
      importedMap.get(s)!.add(cacheKey)
    }
  }

  return as
}
