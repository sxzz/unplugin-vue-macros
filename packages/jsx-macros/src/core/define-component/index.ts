import { importHelperFn, type MagicStringAST } from '@vue-macros/common'
import { walkIdentifiers } from '@vue/compiler-core'
import { camelize } from '@vue/shared'
import type { FunctionalNode, RootMapValue } from '..'
import { restructure } from './restructure'

function getWalkedIds(root: FunctionalNode, propsName: string) {
  const walkedIds = new Set<string>()
  walkIdentifiers(
    root.body,
    (id, parent) => {
      if (
        id.name === propsName &&
        (parent?.type === 'MemberExpression' ||
          parent?.type === 'OptionalMemberExpression')
      ) {
        const prop =
          parent.property.type === 'Identifier'
            ? parent.property.name
            : parent.property.type === 'StringLiteral'
              ? parent.property.value
              : ''
        if (prop) walkedIds.add(prop)
      }
    },
    false,
  )
  return walkedIds
}

export function transformDefineComponent(
  root: FunctionalNode,
  propsName: string,
  map: RootMapValue,
  s: MagicStringAST,
  lib: string,
): void {
  if (!map.defineComponent) return
  s.overwriteNode(
    map.defineComponent.callee,
    importHelperFn(s, 0, 'defineComponent', 'vue'),
  )

  const props: Record<string, string | null> = {}
  if (root.params[0]) {
    if (root.params[0].type === 'Identifier') {
      getWalkedIds(root, propsName).forEach((id) => (props[id] = null))
    } else {
      for (const prop of restructure(s, root)) {
        if (prop.path.endsWith('props')) {
          if (prop.isRest) {
            getWalkedIds(root, prop.name).forEach((id) => (props[id] = null))
          } else {
            props[prop.name] = prop.isRequired ? '{ required: true }' : null
          }
        }
      }
    }
  }

  for (const { expression, modelName, isRequired } of map.defineModel || []) {
    const name = camelize(
      expression.arguments[0]?.type === 'StringLiteral'
        ? expression.arguments[0].value
        : modelName,
    )
    const modelOptions =
      expression.arguments[0]?.type === 'ObjectExpression'
        ? expression.arguments[0]
        : expression.arguments[1]?.type === 'ObjectExpression'
          ? expression.arguments[1]
          : undefined
    const options: any = {}
    if (isRequired) options.required = true
    for (const prop of modelOptions?.properties || []) {
      if (
        prop.type === 'ObjectProperty' &&
        prop.key.type === 'Identifier' &&
        ['validator', 'type', 'required'].includes(prop.key.name)
      ) {
        options[prop.key.name] = s.sliceNode(prop.value)
      }
    }
    props[name] = Object.keys(options).length
      ? `{ ${Object.entries(options)
          .map(([key, value]) => `${key}: ${value}`)
          .join(', ')} }`
      : null
    props[`'onUpdate:${name}'`] = null
  }

  const propsString = Object.entries(props)
    .map(([key, value]) => `${key}: ${value}`)
    .join(', ')
  if (propsString) {
    const argument = map.defineComponent.arguments[1]
    if (!argument) {
      s.appendRight(root.end!, `, { props: { ${propsString} } }`)
    } else if (
      argument.type === 'ObjectExpression' &&
      !argument.properties?.find(
        (prop) =>
          prop.type === 'ObjectProperty' &&
          prop.key.type === 'Identifier' &&
          prop.key.name === 'props',
      )
    ) {
      s.appendLeft(
        argument.end! - 1,
        `${
          !argument.extra?.trailingComma && argument.properties.length
            ? ','
            : ''
        } props: { ${propsString} }`,
      )
    }
  }

  // Auto add `() => ` for return statement
  if (lib === 'vue') {
    if (root.body.type === 'BlockStatement') {
      const returnStatement = root.body.body.find(
        (node) => node.type === 'ReturnStatement',
      )
      if (
        returnStatement &&
        returnStatement.argument &&
        !(
          returnStatement.argument?.type === 'ArrowFunctionExpression' ||
          returnStatement.argument?.type === 'FunctionExpression'
        )
      ) {
        s.appendRight(
          returnStatement.argument.extra?.parenthesized
            ? (returnStatement.argument.extra.parenStart as number)
            : returnStatement.argument.start!,
          '() => ',
        )
      }
    } else {
      s.appendRight(
        root.body.extra?.parenthesized
          ? (root.body.extra.parenStart as number)
          : root.body.start!,
        '() => ',
      )
    }
  }
}
