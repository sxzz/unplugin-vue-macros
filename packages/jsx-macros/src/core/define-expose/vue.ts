import { importHelperFn, type MagicStringAST } from '@vue-macros/common'
import type { CallExpression } from '@babel/types'

export function transformVueDefineExpose(
  node: CallExpression,
  s: MagicStringAST,
  lib: string,
): void {
  s.overwriteNode(
    node.callee,
    importHelperFn(s, 0, 'useExpose', '@vue-macros/jsx-macros/helpers'),
  )
  s.appendRight(
    node.arguments[0]?.start || node.end! - 1,
    `${importHelperFn(s, 0, 'getCurrentInstance', lib)}(), `,
  )
}
