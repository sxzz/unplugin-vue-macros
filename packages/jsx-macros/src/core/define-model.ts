import { importHelperFn, type MagicStringAST } from '@vue-macros/common'
import type { OptionsResolved } from '..'
import { useModelHelperId } from './helper'
import type { CallExpression } from '@babel/types'

export function transformDefineModel(
  node: CallExpression,
  propsName: string,
  s: MagicStringAST,
  options: OptionsResolved,
): void {
  s.overwriteNode(
    node.callee,
    importHelperFn(
      s,
      0,
      'useModel',
      useModelHelperId,
      false,
      options.defineModel.alias[0],
    ),
  )
  s.appendRight(
    node.arguments[0]?.start || node.end! - 1,
    `${propsName}, ${
      node.arguments[0]?.type !== 'StringLiteral' ? `'modelValue',` : ''
    }`,
  )
}
