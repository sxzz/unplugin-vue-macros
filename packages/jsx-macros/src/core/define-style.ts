import { walkAST, type MagicStringAST } from '@vue-macros/common'
import hash from 'hash-sum'
import { helperPrefix } from './helper'
import { isFunctionalNode, type FunctionalNode } from '.'
import type { CallExpression, Node } from '@babel/types'

export function transformDefineStyle(
  node: CallExpression,
  lang: string,
  root: FunctionalNode | undefined,
  defineStyleIndex: number,
  s: MagicStringAST,
  importMap: Map<string, string>,
): void {
  if (node.arguments[0]?.type !== 'TemplateLiteral') return

  let css = s.sliceNode(node.arguments[0]).slice(1, -1)
  const scopeId = hash(css)
  const vars = new Map<string, string>()
  node.arguments[0].expressions.forEach((expression) => {
    const cssVar = s.sliceNode(expression)
    const cssVarId = toCssVarId(cssVar, `--${scopeId}-`)
    s.overwrite(expression.start! - 2, expression.end! + 1, `var(${cssVarId})`)
    vars.set(cssVarId, cssVar)
  })

  let returnExpression = root && getReturnStatement(root)
  if (isFunctionalNode(returnExpression)) {
    returnExpression = getReturnStatement(returnExpression)
  }
  if (vars.size && returnExpression) {
    const children =
      returnExpression.type === 'JSXElement'
        ? [returnExpression]
        : returnExpression.type === 'JSXFragment'
          ? returnExpression.children
          : []
    const varString = Array.from(vars.entries())
      .map(([key, value]) => `'${key}': ${value}`)
      .join(', ')
    for (const child of children) {
      if (child.type === 'JSXElement') {
        s.appendRight(
          child.openingElement.name.end!,
          ` {...{style:{${varString}}}}`,
        )
      }
    }
  }

  let scoped = !!root
  if (node.arguments[1]?.type === 'ObjectExpression') {
    for (const prop of node.arguments[1].properties) {
      if (
        prop.type === 'ObjectProperty' &&
        prop.key.type === 'Identifier' &&
        prop.key.name === 'scoped' &&
        prop.value.type === 'BooleanLiteral'
      ) {
        scoped = prop.value.value
      }
    }
  }

  if (scoped && returnExpression) {
    walkAST<Node>(returnExpression, {
      enter(node) {
        if (
          node.type === 'JSXElement' &&
          s.sliceNode(node.openingElement.name) !== 'template'
        ) {
          s.appendRight(node.openingElement.name.end!, ` data-v-${scopeId}=""`)
        }
      },
    })
  }

  css = s.sliceNode(node.arguments[0]).slice(1, -1)
  const importId = `${helperPrefix}/define-style?index=${defineStyleIndex}&scopeId=${scopeId}&scoped=${scoped}&lang.${lang}`
  importMap.set(importId, css)
  s.appendLeft(0, `import "${importId}";`)
  s.removeNode(node)
}

function getReturnStatement(root: FunctionalNode) {
  if (root.body.type === 'BlockStatement') {
    const returnStatement = root.body.body.find(
      (node) => node.type === 'ReturnStatement',
    )
    if (returnStatement) {
      return returnStatement.argument
    }
  } else {
    return root.body
  }
}

function toCssVarId(name: string, prefix = '') {
  return (
    prefix +
    name.replaceAll(/\W/g, (searchValue, replaceValue) => {
      return searchValue === '.'
        ? '-'
        : name.charCodeAt(replaceValue).toString()
    })
  )
}
