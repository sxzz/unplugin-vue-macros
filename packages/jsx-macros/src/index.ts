import {
  createFilter,
  detectVueVersion,
  FilterFileType,
  getFilterPattern,
  normalizePath,
  REGEX_SETUP_SFC,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import {
  createUnplugin,
  type UnpluginContextMeta,
  type UnpluginInstance,
} from 'unplugin'
import { generatePluginName } from '#macros' with { type: 'macro' }
import { transformJsxMacros } from './core'
import {
  helperPrefix,
  useExposeHelperCode,
  useExposeHelperId,
  useModelHelperCode,
  useModelHelperId,
  withDefaultsCode,
  withDefaultsHelperId,
} from './core/helper'
import { transformStyle } from './core/style'

export type Options = BaseOptions & {
  lib?: 'vue' | 'vue/vapor' | 'react' | 'preact'
  defineStyle?: {
    lang?:
      | 'css'
      | 'postcss'
      | 'scss'
      | 'sass'
      | 'less'
      | 'stylus'
      | (string & {})
    scoped?: boolean
  }
}
export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'lib' | 'defineStyle'
> & { importMap: Map<string, string> }

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const lib = options.lib || 'vue'
  const include = getFilterPattern([FilterFileType.SRC_FILE], framework)
  const defineStyle = options.defineStyle || {}
  defineStyle.lang ??= 'css'
  return {
    include,
    exclude: [REGEX_SETUP_SFC],
    ...options,
    importMap: new Map(),
    defineStyle,
    version,
    lib,
  }
}

const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, true> = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    return [
      {
        name,
        enforce: 'pre',

        resolveId(id) {
          if (normalizePath(id).startsWith(helperPrefix)) return id
        },
        loadInclude(id) {
          return normalizePath(id).startsWith(helperPrefix)
        },
        load(_id) {
          const id = normalizePath(_id)
          if (id === useExposeHelperId) return useExposeHelperCode
          else if (id === useModelHelperId) return useModelHelperCode
          else if (id === withDefaultsHelperId) return withDefaultsCode
        },

        transformInclude: filter,
        transform(code, id) {
          if (options.importMap.get(id)) return
          return transformJsxMacros(code, id, options)
        },
      },
      {
        name: 'unplugin-define-style',
        transformInclude: createFilter({
          include: [new RegExp(`^${helperPrefix}/define-style`)],
        }),
        loadInclude(id) {
          return normalizePath(id).startsWith(helperPrefix)
        },
        load(_id) {
          const id = normalizePath(_id)
          if (options.importMap.get(id)) return options.importMap.get(id)
        },
        transform(code: string, id: string) {
          if (options.importMap.get(id))
            return transformStyle(code, id, options)
        },
      },
    ]
  },
)
export default plugin
