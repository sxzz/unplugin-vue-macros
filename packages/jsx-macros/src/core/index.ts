import {
  babelParse,
  generateTransform,
  getLang,
  HELPER_PREFIX,
  MagicStringAST,
  walkAST,
  type CodeTransform,
} from '@vue-macros/common'
import type { OptionsResolved } from '..'
import { transformDefineComponent } from './define-component'
import { transformDefineExpose } from './define-expose'
import { transformDefineModel } from './define-model'
import { transformDefineSlots } from './define-slots'
import type {
  ArrowFunctionExpression,
  CallExpression,
  FunctionDeclaration,
  FunctionExpression,
  Node,
} from '@babel/types'

export type FunctionalNode =
  | FunctionDeclaration
  | FunctionExpression
  | ArrowFunctionExpression

function getMacroExpression(node?: Node | null) {
  if (node?.type === 'TSNonNullExpression') {
    node = node.expression
  }
  return (
    node?.type === 'CallExpression' &&
    node.callee.type === 'Identifier' &&
    ['defineSlots', 'defineModel', 'defineExpose'].includes(node.callee.name) &&
    node
  )
}

export type RootMapValue = {
  defineComponent?: CallExpression
  defineModel?: {
    expression: CallExpression
    isRequired: boolean
  }[]
  defineSlots?: CallExpression
  defineExpose?: CallExpression
}

function getRootMap(s: MagicStringAST, id: string) {
  const parents: (Node | undefined | null)[] = []
  const rootMap = new Map<FunctionalNode, RootMapValue>()
  walkAST<Node>(babelParse(s.original, getLang(id)), {
    enter(node, parent) {
      parents.unshift(parent)
      const root =
        parents[1] &&
        (parents[1].type === 'ArrowFunctionExpression' ||
          parents[1].type === 'FunctionDeclaration' ||
          parents[1].type === 'FunctionExpression')
          ? parents[1]
          : undefined
      if (!root) return

      if (
        parents[2]?.type === 'CallExpression' &&
        s.sliceNode(parents[2].callee) === 'defineComponent'
      ) {
        if (!rootMap.has(root)) rootMap.set(root, {})
        if (!rootMap.get(root)!.defineComponent) {
          rootMap.get(root)!.defineComponent = parents[2]
        }
      }

      const expression =
        node.type === 'VariableDeclaration'
          ? node.declarations[0].init
          : node.type === 'ExpressionStatement'
            ? node.expression
            : undefined
      const macroExpression = getMacroExpression(expression)
      if (!macroExpression) return
      if (!rootMap.has(root)) rootMap.set(root, {})
      const macroName = s.sliceNode(macroExpression.callee)
      if (macroName) {
        if (macroName === 'defineModel') {
          ;(rootMap.get(root)!.defineModel ??= []).push({
            expression: macroExpression,
            isRequired: expression?.type === 'TSNonNullExpression',
          })
        } else if (
          macroName === 'defineSlots' ||
          macroName === 'defineExpose'
        ) {
          rootMap.get(root)![macroName] = macroExpression
        }
      }
    },
    leave() {
      parents.shift()
    },
  })
  return rootMap
}

export function getParamsStart(node: FunctionalNode, code: string): number {
  return node.params[0]
    ? node.params[0].start!
    : node.start! +
        (code.slice(node.start!, node.body.start!).match(/\(\s*\)/)?.index ||
          0) +
        1
}

export function transformJsxMacros(
  code: string,
  id: string,
  options: OptionsResolved,
): CodeTransform | undefined {
  const s = new MagicStringAST(code)
  const rootMap = getRootMap(s, id)

  for (const [root, map] of rootMap) {
    let propsName = `${HELPER_PREFIX}props`
    if (root.params[0]) {
      if (root.params[0].type === 'Identifier') {
        propsName = root.params[0].name
      } else if (root.params[0].type === 'ObjectPattern') {
        const lastProp = root.params[0].properties.at(-1)
        if (lastProp?.type !== 'RestElement') {
          s.appendRight(
            root.params[0].extra?.trailingComma
              ? (root.params[0].extra?.trailingComma as number) + 1
              : lastProp?.end || root.params[0].end! - 1,
            `${
              !root.params[0].extra?.trailingComma &&
              root.params[0].properties.length
                ? ','
                : ''
            } ...${HELPER_PREFIX}props`,
          )
        }
      }
    } else {
      s.appendRight(getParamsStart(root, s.original), propsName)
    }

    if (map.defineComponent) {
      transformDefineComponent(root, propsName, map, s, options.lib)
    }
    if (map.defineModel?.length) {
      map.defineModel.forEach(({ expression }) => {
        transformDefineModel(expression, propsName, s)
      })
    }
    if (map.defineSlots) {
      transformDefineSlots(map.defineSlots, propsName, s, options.lib)
    }
    if (map.defineExpose) {
      transformDefineExpose(map.defineExpose, root, s, options.lib)
    }
  }

  return generateTransform(s, id)
}
