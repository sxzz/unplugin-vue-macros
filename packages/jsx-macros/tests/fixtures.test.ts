import { testFixtures } from '@vue-macros/test-utils'
import { describe } from 'vitest'
import { transformJsxMacros } from '../src/core'

describe('fixtures', async () => {
  await testFixtures(
    import.meta.glob('./fixtures/**/*.tsx', {
      eager: true,
      as: 'raw',
    }),
    (args, id, code) =>
      transformJsxMacros(code, id, new Map(), {
        lib: 'vue',
        include: ['*.tsx'],
        version: 3.5,
        macros: {
          defineModel: ['defineModel'],
          defineSlots: ['defineSlots'],
          defineExpose: ['defineExpose'],
          defineStyle: ['defineStyle'],
        },
      })?.code,
  )
})

describe('react fixtures', async () => {
  await testFixtures(
    import.meta.glob('./fixtures/**/define-expose.tsx', {
      eager: true,
      as: 'raw',
    }),
    (args, id, code) =>
      transformJsxMacros(code, id, new Map(), {
        lib: 'react',
        include: ['*.tsx'],
        version: 18,
        macros: {
          defineModel: ['defineModel'],
          defineSlots: ['defineSlots'],
          defineExpose: ['defineExpose'],
          defineStyle: ['defineStyle'],
        },
      })?.code,
  )
})
