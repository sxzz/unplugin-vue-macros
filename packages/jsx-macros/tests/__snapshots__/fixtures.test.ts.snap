// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > ./fixtures/define-component.tsx 1`] = `
"
import { defineComponent as __MACROS_defineComponent } from "vue";
import { createDefaultPropsProxy as __MACROS_createDefaultPropsProxy } from "/vue-macros/jsx-macros/with-defaults";
import { createPropsRestProxy as __MACROS_createPropsRestProxy } from "vue";
import { useModel as __MACROS_useModel } from "/vue-macros/jsx-macros/use-model";import { defineComponent } from 'vue'

const Comp = __MACROS_defineComponent(
  (__MACROS_props) => {
const __MACROS_defaults_props = __MACROS_createDefaultPropsProxy(__MACROS_props, {'.bar': 'bar'!});
const props = __MACROS_createPropsRestProxy(__MACROS_defaults_props, ['bar']);
    const foo = __MACROS_useModel(props, 'foo', {
      validator: (value) => {
        return value === 'foo'
      },
      required: false,
      type: String,
    })
    return () => <div>{[foo.value, __MACROS_defaults_props.bar, props.baz]}</div>
  },
  { inheritAttrs: false , props: { bar: { required: true }, baz: null, foo: { validator: (value) => {
        return value === 'foo'
      }, required: false, type: String }, 'onUpdate:foo': null }},
)

const Comp1 = __MACROS_defineComponent((props: { bar: 'bar' }) => {
  const foo = __MACROS_useModel(props, 'foo')
  return () => <div>{[foo.value, props['bar']]}</div>
}, { props: { bar: null, foo: null, 'onUpdate:foo': null } })
"
`;

exports[`fixtures > ./fixtures/define-expose.tsx 1`] = `
"
import { useExpose as __MACROS_useExpose } from "/vue-macros/jsx-macros/use-expose";
import { getCurrentInstance as __MACROS_getCurrentInstance } from "vue";export default function (__MACROS_props) {
  __MACROS_useExpose(__MACROS_getCurrentInstance(), {
    foo: 1,
  })
  return <div />
}
"
`;

exports[`fixtures > ./fixtures/define-model.tsx 1`] = `
"
import { useModel as __MACROS_useModel } from "/vue-macros/jsx-macros/use-model";export const Comp = ({ bar, ...__MACROS_props }: { bar: string }) => {
  const foo = __MACROS_useModel(__MACROS_props, 'foo',{ default: bar })!
  return <div>{foo}</div>
}

export default function (__MACROS_props) {
  const modelValue = __MACROS_useModel<string>(__MACROS_props, 'modelValue',)!
  return (
    <Comp v-model:foo={modelValue.value} bar="bar">
      {modelValue.value}
    </Comp>
  )
}
"
`;

exports[`fixtures > ./fixtures/define-slots.tsx 1`] = `
"
import { getCurrentInstance as __MACROS_getCurrentInstance } from "vue";export const Comp = (__MACROS_props) => {
  const slots = Object.assign<{
    default: () => any
  }>({}, __MACROS_getCurrentInstance()?.slots)
  return <div>{slots.default()}</div>
}

export default function (__MACROS_props) {
  const slots = Object.assign({
    default: () => <div>default</div>,
  },__MACROS_getCurrentInstance()?.slots)
  return <div>{slots.default()}</div>
}
"
`;

exports[`react fixtures > ./fixtures/define-expose.tsx 1`] = `
"
import { forwardRef as __MACROS_forwardRef } from "react";
import { useImperativeHandle as __MACROS_useImperativeHandle } from "react";export default __MACROS_forwardRef(function (__MACROS_props, __MACROS_ref) {
  __MACROS_useImperativeHandle(__MACROS_ref, () =>({
    foo: 1,
  }), [])
  return <div />
})
"
`;
